src/main/java/net/pms/util/jna/UTF16StringByReference.java

```diff
@@ -92,11 +92,8 @@ public String getValue() {
 		if (getPointer() == null) {
 			return null;
 		}
-		int length = 0;
-		while (getPointer().getChar(length) != 0) {
-			length += 2;
-		}
-		char[] value = new char[length / 2];
+		int length = (int) getAllocatedSize();
```

Comment:
@valib I'm sorry to say, but this "fix" is actually a bug.

The reason is that these are null-terminated strings (C style), and they will often turn out to be shorter than the allocated size. Typically you reserve the maximum possible length before calling some function, pass the reference and the external library's function fills the referenced memory with however much it needs to return the string. It then terminates with a "null char" and doesn't touch the remaining bytes in the referenced memory, they will contain whatever was there when the memory was reserved, so it can be anything.

When you use the allocated size when reading the value, you will always read everything, including the null terminator and whatever "random" values are there *if* the string is shorter than the allocated size.

It seems like the tests doesn't catch this situation, which isn't ideal, but this is still the case.

I've debugged it a bit (on Fedora), and it seems that the "cause" for the problem is that `Native.WCHAR_SIZE` now suddenly returns 4 instead of 2 (which indicates that the "native" char is UTF-32). I have no idea why this has suddenly changed with the JNA version, but it's not really important. It reveals a "flaw" in the design of this method. `getPointer().getChar()` retrieves the number of bytes dictated by `Native.WCHAR_SIZE`, but this class is UTF-16 specifically which means that the width of a "code point" is 16 bit/2 bytes per definition.

"Luck" has probably caused this to work in the past, fundamentally `getValue()` (and any other methods that use `getChar()` must be rewritten to use another method to read the content.

Reply:
Thanks for the comment. I'll revert it until a better fix can be done

Reply:
@SubJunk I'm not sure if I see any "easy" fix. It's been too long since I wrote this, and I don't remember all the Objective-C stuff.

I do see a problem with the approach itself. `UTF16StringByReference` isn't very helpful, because it doesn't specify endianness. The way it has "worked" until now is that the endianness of the underlying OS has been used implicitly (handled by JNA), but this ends up being vague and error-prone IMO. I've found that the "real" type that I've created this to "match" is Objective-C's `unichar`: https://developer.apple.com/documentation/foundation/unichar?language=objc

It's defined as `typedef unsigned short unichar;`, but I don't really know if `short` always means 16 bits or if it always has the same endianness. It is probably the case that it is fixed, at least the width, but I'm not sure where to get to find precise answers. If the endianness is always the same, the class could be renamed to `UTF16LEStringByReference` or `UTF16BEStringByReference` and if the width is fixed, a byte array could be read, the null-terminator found, and `new String(bytes, charset) could do the rest.

But, doing any such thing without knowing for sure about the width and the endianness is futile.

It seems that this type is only used by one function in `CoreFoundation` (`CFStringGetCharactersPtr`), and that this method isn't currently in use by "our" code. Thus, the easiest fix might just be to delete the `UTF16StringByReference` class and the one method that relies on it, and solve it if this method is needed in the future. I'm not really a fan of such "solutions", but I'm fearing that getting absolute certainty about endianness and width can take a lot of digging. Testing is worthless, as it only reveals what's the case on the system running the test.

Reply:
Ok cool. It does make sense to remove unused code so that's lucky, since I would have no clue how to fix this :)

Reply:
I've looked more into this, and it seems that what has changed is that they have fixed a bug in JNA where the "bounds check" when reading (get) used to operate with 1 byte "char size" in the past (which is often wrong). The set method used the correct width also in the earlier versions though, so "pure luck" is the reason the tests I wrote didn't fail. This method has never actually worked like it should, so I guess it's not a great loss. It would have been revealed as soon as somebody tried to use it to actually read a unicode string from the macOS API.

Reply:
> I've looked more into this, and it seems that what has changed is that they have fixed a bug in JNA where the "bounds check" when reading (get) used to operate with 1 byte "char size" in the past (which is often wrong). The set method used the correct width also in the earlier versions though, so "pure luck" is the reason the tests I wrote didn't fail.

@Nadahar so my fix is bug or not? Honestly I am not the one who can go deep into the JNA code and your implementations I only tested the results and found that the all OSses returns the same allocated size.